doc/reference/stdio.md

```diff
@@ -0,0 +1,843 @@
+  u16 := integer
+```
+
+Writes an usigned 16 bit integer, encoded in network byte order.
+Returns the number of bytes written.
+
+#### BufferedWriter-write-s16
+```scheme
+(BufferedWriter-write-s16 buf s16) -> fixnum
+  buf := BufferedWriter
+  s16 := integer
+```
+
+Writes a signed 16 bit integer, encoded in network byte order.
+Returns the number of bytes written.
+
+#### BufferedWriter-write-u32
+```scheme
+(BufferedWriter-write-u32 buf u32) -> fixnum
+  buf := BufferedWriter
+  u32 := integer
+```
+
+Writes an usigned 32 bit integer, encoded in network byte order.
+Returns the number of bytes written.
+
+#### BufferedWriter-write-s32
+```scheme
+(BufferedWriter-write-s32 buf s32) -> fixnum
+  buf := BufferedWriter
+  s32 := integer
+```
+
+Writes a signed 32 bit integer, encoded in network byte order.
+Returns the number of bytes written.
+
+#### BufferedWriter-write-u64
+```scheme
+(BufferedWriter-write-u64 buf u64) -> fixnum
+  buf := BufferedWriter
+  u64 := integer
+```
+
+Writes an unsigned 64 bit integer, encoded in network byte order.
+Returns the number of bytes written.
+
+#### BufferedWriter-write-s64
+```scheme
+(BufferedWriter-write-s64 buf s64) -> fixnum
+  buf := BufferedWriter
+  s64 := integer
+```
+
+Reads signed 64 bit integer, encoded in network byte order.
+Returns the number of bytes written.
+
+#### BufferedWriter-write-varuint
+```scheme
+(BufferedWriter-write-varuint buf uint) -> fixnum
+  buf  := BufferedWriter
+  uint := integer
+```
+Writes an unsigned integer with variadic encoding.
+Returns the number of bytes written.
+
+#### BufferedWriter-write-varint
+```scheme
+(BufferedWriter-write-varint buf int) -> fixnum
+  buf := BufferedWriter
+  int := integer
+```
+
+Writes signed integer with variadic encoding.
+Returns the number of bytes written.
+
+#### BufferedWriter-write-char
+```scheme
+(BufferedWriter-write-char buf char) -> fixnum
+  buf  := BufferedWriter
+  char := character
+```
+
+Writes a character with utf8 encoding.
+Returns the number of bytes written.
+
+#### BufferedWriter-write-string
+```scheme
+(BufferdWriter-write-string buf str (start 0) (end (string-length str))) -> fixnum
+  buf := BufferedWriter
+  str := string
+  start  := fixnum; 0 <= start < end
+  end    := fixnum; start < end <= (string-length input)
+```
+Writes a string with utf8 encoding.
+Returns the number of bytes written.
+
+#### BufferedWriter-write-line
+```scheme
+(BufferedWriter-write-line buf str (sep #\newline))
+  buf := BufferedWriter
+  str := string
+  sep := char or list of chars
+```
+Writes a string, followed by a separator.
+Returns the number of bytes written
+
+#### BufferedWriter-flush
+```
+(BufferedWriter-flush buf)
+  buf := BufferedWriter
+```
+
+Drains the buffer into the underlying sync.
+
+#### BufferedWriter-reset!
+```scheme
+(BufferedWriter-reset! buf reader)
+  buf    := BufferedWriter
+  reader := Reader
+```
+
+Resets the underlying writer and buffer state, allowing reuse of buffers from a cache.
```

Comment:
I don't understand the semantics of `reset!` for either read or write buffers.

For reader, does that mean "drop data to the end of the buffer, and keep the file position at the same place", basically, same as reading and dropping `(- end start)` bytes?

And for writer, does that mean "drop data since beginning of the buffer, basically undoing the previous batch of writes since then"?

Please document in all the right places.

Reply:
It means close and drop the underlying Reader/Writer, reset internal offsets, and start anew.
I will improve the documentation.

Reply:
Note that the BufferedWriter flushes.